src/hotspot/share/runtime/handshake.cpp

```diff
@@ -465,60 +443,87 @@ bool HandshakeState::possibly_can_process_handshake() {
+  if (current_thread == handshakee) return "self(JavaThread)";
+  if (current_is_requester) {
+    if (current_thread->is_VM_thread()) return "handshaker(VM Thread)";
+    else return "handshaker(JavaThread)";
+  }
+  if (current_thread->is_VM_thread()) return "cooperative(VM Thread)";
+  else return "cooperative(JavaThread)";
+}
 
-  if (!has_specific_operation(is_direct)){
+HandshakeState::ProcessResult HandshakeState::try_process(HandshakeOperation* match_op) {
+  if (!has_operation()) {
     // JT has already cleared its handshake
-    return _no_operation;
+    return HandshakeState::_no_operation;
   }
 
   if (!possibly_can_process_handshake()) {
     // JT is observed in an unsafe state, it must notice the handshake itself
-    return _not_safe;
+    return HandshakeState::_not_safe;
   }
 
-  // Claim the semaphore if there still an operation to be executed.
-  if (!claim_handshake(is_direct)) {
-    return _state_busy;
+  // Claim the mutex if there still an operation to be executed.
+  if (!claim_handshake()) {
+    return HandshakeState::_claim_failed;
   }
 
-  // Check if the handshake operation is the same as the one we meant to execute. The
-  // handshake could have been already processed by the handshakee and a new handshake
-  // by another JavaThread might be in progress.
-  if (is_direct && op != _operation_direct) {
-    _processing_sem.signal();
-    return _no_operation;
-  }
-
-  // If we own the semaphore at this point and while owning the semaphore
+  // If we own the mutex at this point and while owning the mutex
   // can observe a safe state the thread cannot possibly continue without
-  // getting caught by the semaphore.
-  ProcessResult pr = _not_safe;
-  if (can_process_handshake()) {
-    guarantee(!_processing_sem.trywait(), "we should already own the semaphore");
-    log_trace(handshake)("Processing handshake by %s", Thread::current()->is_VM_thread() ? "VMThread" : "Handshaker");
-    _active_handshaker = Thread::current();
-    op->do_handshake(_handshakee);
-    _active_handshaker = NULL;
-    // Disarm after we have executed the operation.
-    clear_handshake(is_direct);
-    pr = _success;
+  // getting caught by the mutex.
+  if (!can_process_handshake()) {
+    _lock.unlock();
+    return HandshakeState::_not_safe;
   }
 
-  // Release the thread
-  _processing_sem.signal();
+  Thread* current_thread = Thread::current();
+
+  HandshakeState::ProcessResult pr_ret = HandshakeState::_processed;
+  int executed = 0;
+
+  do {
+    HandshakeOperation* op = pop_for_processor();
+    if (op != NULL) {
+      assert(SafepointMechanism::local_poll_armed(_handshakee), "Must be");
+      assert(op->_target == NULL || _handshakee == op->_target, "Wrong thread");
+      log_trace(handshake)("Processing handshake " INTPTR_FORMAT " by %s", p2i(op),
+                           executioner_name(current_thread, _handshakee, op == match_op));
```

Comment:
With the above change we could even avoid factoring the code into executioner_name() and just do: 
log_trace(handshake)("Processing handshake " INTPTR_FORMAT " by %s%s", p2i(op),
                                         op == match_op ? "handshaker" : "cooperative", 
                                         current_thread->is_VM_thread() ? "(VM Thread)" : "(JavaThread)");

Reply:
I added a second log line where I use that function also:
  log_trace(handshake)("Thread %s(" INTPTR_FORMAT ") executed %d ops for JavaThread: " INTPTR_FORMAT " %s target op: " INTPTR_FORMAT,
                       executioner_name(current_thread, _handshakee, pr_ret == HandshakeState::_succeed),
                       p2i(current_thread), executed, p2i(_handshakee),
                       pr_ret == HandshakeState::_succeed ? "including" : "excluding", p2i(match_op));

Reply:
Push commit 469f8fc, please have a look.